tools/flow_grid.py

```diff
@@ -0,0 +1,695 @@
+    flowdir : Generate a flow direction grid from a given digital elevation
+              dataset (dem).
+    catchment : Delineate the watershed for a given pour point (x, y)
+                or (column, row).
+    fraction : Generate the fractional contributing area for a coarse
+               scale flow direction grid based on a fine-scale flow
+               direction grid.
+    """
+
+    def __init__(self, **kwargs):
+        self.grid_props = {}
+        if 'data' in kwargs:
+            self.read_input(**kwargs)
+        else:
+            pass
+
+    def read_input(self, data, data_type='dir', input_type='ascii', band=1,
+                   nodata=0, bbox=None, crs=None, **kwargs):
+        """
+        Reads data into a named attribute of flow_grid
+        (name of attribute determined by 'data_type').
+
+        Parameters
+        ----------
+        data : File name (string) or numpy ndarray
+               If data is from a file, 'input_type' should be set to the
+               appropriate value ('ascii' or 'raster').
+        data_type : 'dem', 'dir', 'acc' or other string
+                     Name of dataset. Will determine the name of the attribute
+                     representing the gridded data. Default values are used
+                     internally by some class methods:
+                         'dem' : digital elevation data
+                         'dir' : flow direction data
+                         'acc' : flow accumulation (upstream area) data
+                    
+        input_type : 'raster', 'ascii' or 'array'
+                     Type of input data.
+        band : int
+               For raster data, the band number to read.
+        nodata : int or float
+                 Value indicating no data.
+        bbox : tuple or list
+               Bounding box, if none provided.
+
+        """
+
+        # read ascii file
+        if input_type == 'ascii':
+            with open(data) as header:
+                ncols = int(header.readline().split()[1])
+                nrows = int(header.readline().split()[1])
+                xll = ast.literal_eval(header.readline().split()[1])
+                yll = ast.literal_eval(header.readline().split()[1])
+                cellsize = ast.literal_eval(header.readline().split()[1])
+                nodata = ast.literal_eval(header.readline().split()[1])
+                shape = (nrows, ncols)
+                bbox = (xll, yll, xll + ncols*cellsize, yll + nrows*cellsize)
+            data = np.loadtxt(data, skiprows=6, **kwargs)
+            nodata = data.dtype.type(nodata)
+
+        # read raster file
+        if input_type == 'raster':
+            import rasterio
+            f = rasterio.open(data)
+            crs = f.crs
+            bbox = tuple(f.bounds)
+            shape = f.shape
+            cellsize = f.affine[0]
+            nodata = f.nodatavals[0]
+            if len(f.indexes) > 1:
+                data = np.ma.filled(f.read_band(band))
```

Comment:
In many cases, using numpy masked arrays is quite slow since most of that module is written in pure python. To avoid this, would it work fill with `nan`s?